src/main/scala/interfaces/Converter.scala

```diff
@@ -0,0 +1,318 @@
+                                  First, Second, SortWrapper, PredicateLookup, toSnapTree, Rational, PermLiteral}
+
+
+/* Some new classes to describe a more informative model */
+
+/* Explanation: (to be removed later)
+    To use these new extracted counterexamples, you can use the flag "--counterexample mapped"
+    The new trait ExtractedModelEntry should make the counterexamples more readable and extract
+    information from the heap, but heap values such as predicates can still be accessed,
+    also in a more processed way since all the first/second/Sortwrappers should be gone now and 
+    replaced with their corresponding values in the extracted Heap fields of the converter class
+
+    A lot of types like sequences etc are not handled yet, and I am not sure if I plan to do so
+    since they are not used in Rust
+
+    The various parts in this file should probably be moved out of the interfaces package,
+    but for now to keep everything in one place is simpler to work on it.
+
+    To see the advantage of these new counterexamples you can run it on the
+    files in src/test/resourcers/demo-counterexamples, the file simple-refs-rec.vpr is 
+    probably the most interesting.
+
+*/
+
+sealed trait ExtractedModelEntry
+case class LitIntEntry(value: BigInt) extends ExtractedModelEntry {
+    override def toString = value.toString
+}
+case class LitBoolEntry(value: Boolean) extends ExtractedModelEntry {
+    override def toString = value.toString
+}
+case class RefEntry(fields: Map[String, ExtractedModelEntry]) extends ExtractedModelEntry {
+    override def toString : String = {
+        var buf = "Ref {\n"
+        for ((name, entry) <- fields) {
+            buf += "\t" + name + " <- " + entry.toString.split("\n").mkString("\n\t") + "\n"
+        }
+        buf += "}"
+        return buf
+    }
+}
+case class VarEntry(name: String) {
+    override def toString = name
+}
+case class NullRefEntry() extends ExtractedModelEntry {
+    override def toString = "Null()"
+}
+case class OtherEntry(value: String) extends ExtractedModelEntry {
+    override def toString = value + "(unhandled type)"
+}
+
+case class ExtractedModel(entries: Map[String, ExtractedModelEntry]) {
+    override def toString : String = {
+        var buf = ""
+        for ((name, entry) <- entries) {
+            buf += name + " <- " + entry.toString + "\n"
+        }
+        return buf
+    }
+}
+
+sealed trait HeapEntry
+case class PredHeapEntry(name: String, args: Array[Term]) extends HeapEntry {
+    override def toString = name + "(" + args.mkString(", ") +")"
+}
+case class FieldHeapEntry(recv: String, field: String, perm: Rational, typ: Option[ast.Type]) extends HeapEntry {
+    override def toString = recv + "." + field
+}
+
+case class ExtractedHeap(entries: Map[HeapEntry, String]) {
+    override def toString = entries.map(x => x._1.toString + " <- " + x._2).mkString("\n")
+}
+
+
+/* basically a 1 to 1 copy of nagini code */
+object Converter{    
```

Comment:
Not that Silicon strictly follows best-practice software engineering, but it seems odd that so much code would reside in a file in the `interfaces` directory

Reply:
First of all thanks for the extensive feedback!
I am aware that my current structure does not make a lot of sense, but I thought it would be a good idea to maybe ask you first how you prefer it to be done. Should I put Converter.scala into the reporter package? Or keep the ExtractedHeap and ExtractedModel structure in interfaces and move the rest? I am not very familiar with most of the project.

Reply:
You're welcome! And no worries, I'm aware that this PR is work in progress. Plus, Silicon's structure & architecture are also not totally thought through. For examples, the `interfaces` package is basically an artefact from my Java background, when I began with Silicon; I don't think it's a common separation to have in Scala.

That said: I suggest to keep everything in package `reporter`, potentially in a subpackage thereof, if you add several files.